src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,329 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+namespace Microsoft.Quantum.Simulation.QCTraceSimulatorRuntime {
+
+    /// <summary>
+    /// Either time interval or point in time on the timeline of the program.
+    /// </summary>
+    internal struct ComplexTime
+    {
+        /// <summary>
+        /// Tracer configuration interface uses type "double" to provide duration of gates (aka depth).
+        /// When a gate is performed, time advances by the depth of the gate. DepthTime is used to count
+        /// time cosistent with the depth configuration settings. As a result, if a gate time is set to zero,
+        /// DepthTime won't increase after such gate is performed.
+        /// </summary>
+        internal double DepthTime { get; private set; }
+
+        /// <summary>
+        /// TrailingZeroDepthGateCount counts number of zero-depth gates after last non-zero depth gate is performed.
+        /// Generally, gates that happen later on a qubit cannot be scheduled before gates on the same qubit.
+        /// This is tracked transitively via qubit availability time, but zero-depth gates do not allow such calculation.
+        /// The situation is resolved by having this time component - TrailingZeroDepthGateCount.
+        /// </summary>
+        private long TrailingZeroDepthGateCount;
+
+        internal ComplexTime(double time) {
+            DepthTime = time;
+            TrailingZeroDepthGateCount = 0;
+        }
+
+        private ComplexTime(double depthTime, long trailingZeroDepthGateCount) {
+            DepthTime = depthTime;
+            TrailingZeroDepthGateCount = trailingZeroDepthGateCount;
+        }
+
+        public override string ToString() {
+            return $"{DepthTime}d+{TrailingZeroDepthGateCount}";
+        }
+        /// <summary>
+        /// Add gate time to availability time. If gate time is zero, advance dependency time.
+        /// </summary>
+        /// <param name="gateTime">Gate time to advance by. Assumed to be precise for comparison with 0.</param>
```

Comment:
I believe, you mean here that we expect the gate time to be set from literals rather than to be calculated so the `== 0.0` comparison is safe, right? Maybe, add an assert after the `if` that gateTime is greater than some reasonable epsilon value?

Reply:
Yes, assumed to be set from literals. I updated the comment. I'm not sure what value would be reasonable... A very reasonable restriction would be to restrict these to integers, but for some reason this was made double in the interface.